editor/src/components/canvas/canvas-strategies/strategies/reparent-helpers/reparent-property-strategies.ts

```diff
@@ -187,6 +186,105 @@ export const convertSizingToVisualSizeWhenPastingFromFlexToFlex =
+    }
+
+    const targetParentBounds = MetadataUtils.getFrameInCanvasCoords(targetParent, metadata)
+    if (targetParentBounds == null || isInfinityRectangle(targetParentBounds)) {
+      return left('Target parent bounds are invalid')
+    }
+
+    const elementBounds = MetadataUtils.getFrameInCanvasCoords(elementToReparent.oldPath, metadata)
+    if (elementBounds == null || isInfinityRectangle(elementBounds)) {
+      return left('Element bounds are invalid')
+    }
+
+    const deltaX = elementBounds.x - targetParentBounds.x
+    const deltaY = elementBounds.y - targetParentBounds.y
+
+    const elementInBoundsHorizontally = 0 <= deltaX && deltaX <= targetParentBounds.width
+    const elementInBoundsVertically = 0 <= deltaY && deltaY <= targetParentBounds.height
+
+    const horizontalCenter = roundTo((targetParentBounds.width - elementBounds.width) / 2, 0)
+    const verticalCenter = roundTo((targetParentBounds.height - elementBounds.height) / 2, 0)
+
+    const horizontalOffset = elementInBoundsHorizontally ? deltaX : horizontalCenter
+    const verticalOffset = elementInBoundsVertically ? deltaY : verticalCenter
+
+    return right([
+      ...pruneFlexPropsCommands(flexChildProps, elementToReparent.newPath),
+      setCssLengthProperty(
+        'always',
+        elementToReparent.newPath,
+        PP.create('style', 'top'),
+        { type: 'EXPLICIT_CSS_NUMBER', value: cssNumber(verticalOffset, null) },
+        null,
+      ),
+      setCssLengthProperty(
+        'always',
+        elementToReparent.newPath,
+        PP.create('style', 'left'),
+        { type: 'EXPLICIT_CSS_NUMBER', value: cssNumber(horizontalOffset, null) },
+        null,
+      ),
+      setProperty('always', elementToReparent.newPath, styleP('position'), 'absolute'),
+    ])
+  }
+
+const getZIndex = (element: JSXElement): number | null => {
+  const zIndex = getLayoutProperty('zIndex', right(element.props), styleStringInArray)
+  return foldEither(
+    () => null,
+    (z) => z?.value ?? null,
+    zIndex,
+  )
+}
+
+export const setZIndexOnPastedElement =
+  (
+    elementToReparent: { oldPath: ElementPath; newPath: ElementPath },
+    targetParent: ElementPath,
+    metadata: ElementInstanceMetadataMap,
+  ): ReparentPropertyStrategy =>
+  () => {
+    const siblings = MetadataUtils.getChildrenUnordered(metadata, targetParent)
+    const asd = mapDropNulls((sibling) => {
+      return foldEither(
+        () => null,
+        (e) => (isJSXElement(e) ? getZIndex(e) : null),
+        sibling.element,
+      )
+    }, siblings)
+
+    const maximumZIndexOfOverlappingElements = asd.sort().reverse().at(0)
+
+    if (maximumZIndexOfOverlappingElements == null) {
+      return left('No siblings have z-index applied')
+    }
+
+    return right([
+      setCssLengthProperty(
+        'always',
+        elementToReparent.newPath,
+        PP.create('style', 'zIndex'),
+        setExplicitCssValue(cssNumber(maximumZIndexOfOverlappingElements, null)),
```

Comment:
Should this be +1 to ensure it's on top?

Reply:
I thought about it and I don't this so because of the following:
- When reparenting via copy/paste, we place the pasted element on top, so even if the same z-index is applied, it'll appear on top due to the order of the siblings.
- When reparenting through the navigator, there's a very precise indication of where the element will end up, and even after it's very easy to adjust the position of the element (either via the navigator or the bring forward/send backward shortcuts).

I think this strikes a good 80/20 balance in terms of time spent vs polished behaviour